File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leafs1 = self.leafs(root1)
        leafs2 = self.leafs(root2)
        while True:
            try:
                val1 = next(leafs1)
                done1 = True
            except:
                done1 = False

            try:
                val2 = next(leafs2)
                done2 = True
            except:
                done2 = False

            if done1 and done2:
                if val1 != val2:
                    return False
            elif not done1 and not done2:
                return True
            else:
                return False


    def leafs(self, node):
        if not node.left and not node.right:
            yield node.val
        else:
            if node.left:
                yield from self.leafs(node.left)
            if node.right:
                yield from self.leafs(node.right)

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    const leafIterator1 = new LeafIterator(root1);
        const leafIterator2 = new LeafIterator(root2);
        while (leafIterator1.hasNext() && leafIterator2.hasNext()) {
            if (leafIterator1.next() !== leafIterator2.next()) return false;
        }
        return !(leafIterator1.hasNext() || leafIterator2.hasNext());
};

class LeafIterator {
    constructor(root) {
        this.stack = [];
        while (root !== null) {
            this.stack.push(root);
            root = root.left;
        }
        this.answer = -1;
    }

    hasNext() {
        return this.answer !== null;
    }

    next() {
        while (this.stack.length > 0 && !this.isLeaf(this.stack[this.stack.length - 1])) {
            let node = this.stack.pop().right;
            while (node !== null) {
                this.stack.push(node);
                node = node.left;
            }
        }
        this.answer = this.stack.length === 0 ? null : this.stack.pop().val;
        return this.answer;
    }

    isLeaf(node) {
        return node !== null && node.left === null && node.right === null;
    }
}